torchtext/prototype/t5/modules.py

```diff
@@ -53,8 +54,244 @@ def __init__(
+            if key_padding_mask is not None:
+                key_padding_mask = key_padding_mask.unsqueeze(0)
+
+        # Set up shape vars
+        bsz, tgt_len, embed_dim = query.shape
+        _, src_len, _ = key.shape
+
+        assert (
+            embed_dim == self.embed_dim
+        ), f"was expecting embedding dimension of {self.embed_dim}, but got {embed_dim}"
+        if isinstance(embed_dim, Tensor):
+            # Embed_dim can be a tensor when JIT tracing
+            head_dim = embed_dim.div(self.num_heads, rounding_mode="trunc")
+        else:
+            head_dim = embed_dim // self.num_heads
+        assert (
+            head_dim * self.num_heads == embed_dim
+        ), f"embed_dim {embed_dim} not divisible by num_heads {self.num_heads}"
+        # Allow MHA to have different embedding dimensions when separate projection weights are used
+        assert (
+            key.shape[:2] == value.shape[:2]
+        ), f"key's sequence and batch dims {key.shape[:2]} do not match value's {value.shape[:2]}"
+
+        # Compute in-projection
+        assert self.q_proj_weight is not None, "q_proj_weight is None"
+        assert self.k_proj_weight is not None, "k_proj_weight is None"
+        assert self.v_proj_weight is not None, "v_proj_weight is None"
+        if self.in_proj_bias is None:
+            b_q = b_k = b_v = None
+        else:
+            b_q, b_k, b_v = self.in_proj_bias.chunk(3)
+        q, k, v = F._in_projection(
+            query, key, value, self.q_proj_weight, self.k_proj_weight, self.v_proj_weight, b_q, b_k, b_v
+        )
+
+        # Prep attention mask
+        if attn_mask is not None:
+            if attn_mask.dtype == torch.uint8:
+                warnings.warn("Byte tensor for attn_mask is not supported. Using bool tensor instead.")
+                attn_mask = attn_mask.to(torch.bool)
+            else:
+                assert (
+                    attn_mask.is_floating_point() or attn_mask.dtype == torch.bool
+                ), f"Only float and bool types are supported for attn_mask, not {attn_mask.dtype}"
+            # Ensure attn_mask's dim is 3
+            if attn_mask.dim() == 2:
+                correct_2d_size = (tgt_len, src_len)
+                if attn_mask.shape != correct_2d_size:
+                    raise RuntimeError(
+                        f"The shape of the 2D attn_mask is {attn_mask.shape}, but should be {correct_2d_size}."
+                    )
+                attn_mask = attn_mask.view(1, 1, tgt_len, tgt_len).expand(bsz, self.num_heads, -1, -1)
+            else:
+                raise RuntimeError(f"attn_mask's dimension {attn_mask.dim()} is not supported")
+
+        # Prep key padding mask
+        if key_padding_mask is not None and key_padding_mask.dtype == torch.uint8:
+            warnings.warn("Byte tensor for key_padding_mask is not supported. Using bool tensor instead.")
+            key_padding_mask = key_padding_mask.to(torch.bool)
+
```

Comment:
We were having issues from the original implementation where the q, k, v tensors were reshaped to `(batch_size * num_heads, seq_len, head_dim)`. This led to outputs that differed from the HF output when the input sequence to the decoder had batch size larger 1. The resolution to this was to shape these into 4D tensors of shape `(batch_size, num_heads, seq_len, head_dim)` which is the same reshaping done in the HF implementation.